NeoML/src/TraditionalML/SparseFloatMatrix.cpp

```diff
@@ -158,32 +160,39 @@ void CSparseFloatMatrix::AddRow( const CSparseFloatVectorDesc& row )
 
 	CSparseFloatMatrixBody* newBody = body.CopyOnWrite();
 	newBody->Desc.Height++;
-	newBody->Desc.Width = max( body->Desc.Width, row.Size == 0 ? 0 : row.Indexes[row.Size - 1] + 1 );
 	newBody->Desc.PointerB[newBody->Desc.Height - 1] = newBody->ElementCount;
-	newBody->Desc.PointerE[newBody->Desc.Height - 1] = newBody->ElementCount + row.Size;
-	::memcpy( newBody->Desc.Columns + newBody->ElementCount, row.Indexes, row.Size * sizeof( int ) );
-	::memcpy( newBody->Desc.Values + newBody->ElementCount, row.Values, row.Size * sizeof( float ) );
-	newBody->ElementCount += row.Size;
+	int size = row.Size;
+	int* indexes = newBody->Desc.Columns + newBody->ElementCount;
+	float* values = newBody->Desc.Values + newBody->ElementCount;
+	if( row.Indexes == nullptr && row.Values != nullptr ) {
+		int k = 0;
+		for( int i = 0; i < row.Size; ++i ) {
+			if( row.Values[i] != 0 ) {
+				indexes[k] = i;
+				values[k] = row.Values[i];
+				++k;
+			}
+		}
+		size = k;
+	} else {
+		::memcpy( indexes, row.Indexes, row.Size * sizeof( int ) );
+		::memcpy( values, row.Values, row.Size * sizeof( float ) );
+	}
+	newBody->Desc.Width = max( body->Desc.Width, size == 0 ? 0 : indexes[size - 1] + 1 );
+	newBody->Desc.PointerE[newBody->Desc.Height - 1] = newBody->ElementCount + size;
+	newBody->ElementCount += size;
 }
 
 CSparseFloatVectorDesc CSparseFloatMatrix::GetRow( int index ) const
 {
 	NeoAssert( 0 <= index && index < GetHeight() );
-
-	CSparseFloatVectorDesc res;
-	res.Size = body == 0 ? 0 : body->Desc.PointerE[index] - body->Desc.PointerB[index];
-	res.Indexes = body == 0 ? 0 : body->Desc.Columns + body->Desc.PointerB[index];
-	res.Values = body == 0 ? 0 : body->Desc.Values + body->Desc.PointerB[index];
-	return res;
+	return body->Desc.GetRow( index );
```

Comment:
body == 0 ?

Reply:
GetHeight() in NeoAssert returns 0 if body == 0. Combining with the first condition `index >= 0` index can't be lower then 0 so body == nullptr is impossible here

Reply:
Add NeoAssert( body != nullptr );